ethereumj-core/src/main/java/org/ethereum/sharding/manager/ShardingWorldManager.java

```diff
@@ -0,0 +1,128 @@
+import org.ethereum.core.TransactionExecutor;
+import org.ethereum.crypto.ECKey;
+import org.ethereum.db.BlockStore;
+import org.ethereum.db.DbFlushManager;
+import org.ethereum.db.RepositoryWrapper;
+import org.ethereum.listener.EthereumListener;
+import org.ethereum.listener.EthereumListenerAdapter;
+import org.ethereum.manager.WorldManager;
+import org.ethereum.sharding.config.DepositContractConfig;
+import org.ethereum.sharding.service.ValidatorService;
+import org.ethereum.vm.program.invoke.ProgramInvokeFactoryImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.spongycastle.util.encoders.Hex;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.math.BigInteger;
+import java.util.concurrent.CompletableFuture;
+
+import static org.ethereum.util.ByteUtil.longToBytes;
+
+/**
+ * In addition to original {@link WorldManager} runs sharding bootstrap things:
+ * <ul>
+ *     <li> deploys validator registration contract to the empty blockchain
+ *     <li> starts {@link ValidatorService} if it's enabled
+ * </ul>
+ *
+ * @author Mikhail Kalinin
+ * @since 26.07.2018
+ */
+public class ShardingWorldManager extends WorldManager {
+
+    private static final Logger logger = LoggerFactory.getLogger("sharding");
+
+    private static final ECKey DEPLOY_AUTHORITY = ECKey.fromPrivate(Hex.decode("41791102999c339c844880b23950704cc43aa840f3739e365323cda4dfa89e7a"));
+    private static final long DEPLOY_GAS_PRICE = 1_000_000_000; // 1 GWEI
+    private static final long DEPLOY_GAS_LIMIT = 200_000;
+
+    DepositContractConfig contractConfig;
+    DbFlushManager dbFlushManager;
+
+    private CompletableFuture<Void> contractInit = new CompletableFuture<>();
+
+    public ShardingWorldManager(SystemProperties config, Repository repository, EthereumListener listener,
+                                Blockchain blockchain, BlockStore blockStore, DepositContractConfig contractConfig,
+                                DbFlushManager dbFlushManager) {
+        super(config, repository, listener, blockchain, blockStore);
+        this.contractConfig = contractConfig;
+        this.dbFlushManager = dbFlushManager;
+    }
+
+    @Override
+    protected void init() {
+        initDepositContract();
+        syncManager.init(getChannelManager(), pool);
+    }
+
+    private void initDepositContract() {
+        if (getBlockchain().getBestBlock().isGenesis()) {
+            Block best = getBlockchain().getBestBlock();
+
+            // build and execute deploy transaction
+            BigInteger nonce = getRepository().getNonce(DEPLOY_AUTHORITY.getAddress());
+            Integer chainId = config.getBlockchainConfig().getConfigForBlock(0).getChainId();
+            Transaction tx = new Transaction(nonce.toByteArray(), longToBytes(DEPLOY_GAS_PRICE), longToBytes(DEPLOY_GAS_LIMIT),
+                    null, BigInteger.ZERO.toByteArray(), contractConfig.getBin(), chainId) {
+                @Override
```

Comment:
Any alternatives to this kind of hack?
I guess you want it to be catched by logfilter, yeah

Reply:
This is a hack indeed.
It was not enough to just save the code of contract to predefined address. It also requires to run the code which does initial set up of the contract. Thus it needs to be executed as a regular tx but also stored to some certain address and I just haven't found anything less intrusive than address overriding.

Reply:
Ok, I got it.